argilla/src/argilla/records/_dataset_records.py

```diff
@@ -494,3 +517,52 @@ def _infer_record_from_mapping(
             responses=responses,
             _dataset=self.__dataset,
         )
+
+    def _parse_suggestion_from_mapping(
+        self, data: Dict[str, Any], mapping: Dict[str, Tuple[Optional[str]]]
+    ) -> List[Suggestion]:
+        """Converts a mapped suggestion dictionary to a Suggestion object for use by the add or update methods.
+        Suggestions can be defined across multiple source values and mapped to single questions.
+        Args:
+            data: A dictionary representing the suggestion.
+            mapping: A dictionary mapping from source data keys/ columns to Argilla fields, questions, ids, etc.
+        Returns:
+            A Suggestion object.
+
+        """
+        suggestion_values = defaultdict(dict)
+
+        for (attribute_name, sub_attribute), source_key in mapping.items():
+            value = data.get(source_key)
+            schema_item = self.__dataset.schema.get(attribute_name)
+            suggestion_values[attribute_name].update(
+                {sub_attribute: value, "question_name": attribute_name, "question_id": schema_item.id}
```

Comment:
```suggestion
                {sub_attribute: value, "question_name": attribute_name, "question_id": schema_item.id}
```
maybe `sub_attribute` is not very specific and can be `attribute_value_key` or something?

Reply:
sub_attribute as yet only refers to the parameters of a suggestion. So score and agent. I don't think `_key` explains that, but I'm open to suggestions.

Reply:
Okay. Now I think about what these methods do. So the idea is to generate a normalized mapping which will be used to build the record. Maybe this mapping can be done once and used for all the records since we have the dataset schema which can be used as a default mapping. I have some comments about the implementation.

1. The `_render_record_mapping` can be executed  just once if we use the settings schema
2. I think it would be easier to follow if the "singular_mapping" is a class instead of a raw dict. This will be a normalized representation of the mapping so we can have something like:

```python
# this will apply the `_render_record_mapping` logic (I guess)
mapping = NormalizedMapping(mapping, schema=settings.schema)

# at some point inside _infer_record_from_mapping
# Parse suggestions and responses into objects aligned with questions
suggestions = self._parse_suggestion_from_mapping(data=data, mapping=mapping.suggestions)
responses = self._parse_response_from_mapping(data=data, mapping=mapping.responses, user_id=user_id)

suggestions = [Suggestion(**suggestion_dict) for suggestion_dict in suggestion_values.values()]
# Parse fields, metadata, and vectors into
fields = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.fields.items()}
metadata = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.metadata.items()}
vectors = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.vectors.items()}
```

or even more simple:
```python
mapping = NormalizedMapping(mapping, schema=settings.schema)
...

suggestions = mapping.suggestions(data)
responses = mapping.responses(data, user_id)
fields = mapping.fields(data)
metadata = mapping.metadata(data)
vectors = mapping.vectors(data)

```
or just in once:
```python
record = mapping.infer_record(data)
```

Reply:
@frascuchon I've reimplemented the PR using a class as suggested.